### DISCUSSION TITLE
How to crop an image on a bounding box?

### DISCUSSION THREAD

**Discussion by kj7rrv** [ImageMagick/ImageMagick#5443] [Digital Image Processing] [answered]
I need to crop an image on a bounding box. It's a 2020 USGS topo map from [topoView](https://ngmdb.usgs.gov/topoview/viewer/#11/42.0462/-124.3244). [
![small](https://user-images.githubusercontent.com/58444735/184553045-b31f3483-6146-467a-b12a-8140129f441c.jpg)
This](https://github.com/ImageMagick/ImageMagick/files/9334030/OR_Brookings_20200316_TM_geo_jpg.zip) (higher-resolution version of the above; JPEG in ZIP file) is an example, but it needs to work on any of the 2020 maps. Is there a way to do that with ImageMagick?

**Comment by fmw42** [comment]
JPG is a lousy format to use for extracting that area. Solid color are not actually solid everywhere, especially near your number and letters. So thresholding will not result in clean images to find the largest rectangle. Do you have access to non-lossy compressed format (not converted from JPG)?

**Reply by kj7rrv** [reply]
@fmw42 TIFF is available. AFAIK it isn't converted from JPEG.

**Reply by fmw42** [reply]
Zip and attach one example TIFF

**Reply by fmw42** [reply]
What is your IM version and platform?

**Reply by kj7rrv** [reply]
https://ngmdb.usgs.gov/ht-bin/tv_download.pl?id=0be35c89341a750ed6ad414f679353aa&fmt=tif (I'll be working with OR_Brookings_20200316_TM_geo.tif). ImageMagick 7.1.0-45 on Arch Linux

**Reply by fmw42** [reply]
Link is bad for me. Page cannot be found.  Do you need login access?

**Reply by kj7rrv** [reply]
@fmw42 hmm, it's 404ing for me now too. Try https://kj7rrv.com/OR_Brookings_20200316_TM_geo.tif I uploaded that to my Web server; it should work.

**Reply by fmw42** [reply]
That TIF is JPG compressed. So the same issue.  Do you have access to non-JPG compressed TIFF?

**Reply by kj7rrv** [reply]
@fmw42 no; KMZ https://kj7rrv.com/OR_Brookings_20200316_TM_geo_kmz.zip and GeoPDF https://kj7rrv.com/OR_Brookings_20200316_TM_geo.pdf are available; IDK if they're JPG-compressed as well. I don't like the PDF one because there are too many contour lines, but I can use it if it isn't possible to use the JPEG or TIFF.

If I have to use the JPEG or TIFF, there's no benefit in using the TIFF because it's also JPEG-compressed, right?

**Comment by fmw42** [comment]
I do not know if this will help, but you can use connected components processing to find the largest region in a thresholded  version of the image. Then get the bounding box and crop the original at the same locations. Unfortunately, due to the JPG compression, I cannot separate the map region from the grid number on the sides. And I can only do that at the third resolution level (IM counts from 0, so number 2). This is because when thresholding, I get too many regions error message.  I converted the image to PNG levels, but that can be avoided by using the tiff layer numbers.

```
bbox=$(magick OR_Brookings_20200316_TM_geo-2.png -auto-level -negate -threshold 2% -type bilevel \
-define connected-components:verbose=true \
-define connected-components:mean-color=true \
-define connected-components:keep-top=1 \
-connected-components 4 x.png | grep "gray(255)" | awk '{print $2}')
echo $bbox
magick OR_Brookings_20200316_TM_geo-2.png -crop "$bbox" +repage OR_Brookings_20200316_TM_geo-2_crop.png
```

![OR_Brookings_20200316_TM_geo-2_crop](https://user-images.githubusercontent.com/13458218/184558773-54eb60b1-ad24-496a-95ea-924427099ef0.png)


To get the crop from the highest resolution, just scale up the bounding box values by a factor of 8 and use those on the full resolution layer.

**Reply by kj7rrv** [reply]
Thank you! Would it be possible to reduce the color depth and remove the compression artifacts that way?

**Reply by fmw42** [reply]
I reduced the depth to 8 before processing when I converted to PNG. The processing above then thresholds to binary. You can play with the threshold value in the command. But JPG compression causes the number to blur when thresholding.

Sorry, I do not know what else to try. Perhaps someone else on the forum might be able help further.

**Reply by kj7rrv** [reply]
Okay, thank you for the help. This is definitely a good start; I'll see what I can do from here.

**Reply by kj7rrv** [reply]
When I try to run the code on a PNG converted from JPEG, I get "too many objects". Is there a way to fix that?

**Reply by fmw42** [reply]
FYI, I was trying to threshold so that I selected the white and made all the rest black (then inverted that). But look at the white region around the numbers on the sides of the map. You will see that they are not pure white like the rest of the white background. This is due to JPG compression. That causes me to have to increase the threshold value from 0 to 2%.

**Reply by kj7rrv** [reply]
Yeah, I see that. What happens if you set an even higher threshold? I would try it, but it just gives the "too many objects" error on my computer.

**Reply by fmw42** [reply]
That is what I said. You must use a lower resolution such as level 3 or 4 to avoid that. You may be able to do better on Q32 compile of IM. But for me that did not help with this image.  The highest level that you can use may also be image dependent. Also the threshold value may be image dependent. So try other thresholds.

**Reply by kj7rrv** [reply]
Oh, okay. Could I scale the picture down, then scale the bounding box up and crop the full-resolution image? I need the final result to be full-resolution.

**Reply by fmw42** [reply]
That is what I said in my answer. You do not need to scale the image down. Just use one of the lower resolution layers from the TIFF (i.e. around 3). Use whatever level does not give the too many regions error.  Then depending upon the resolution level, scale the bounding box values. So if you are using level 3 that would be a factor of 8 (2*2*2), assuming each level is a factor of 2 lower. See my answer above with the code.

**Comment by snibgo** [comment]
I don't understand what is wanted. "Crop an image on a bounding box", but what bounding box?

Perhaps a sample required output would help.

**Reply by fmw42** [reply]
I believe that he just wants the map portion of the image bounded by the black line all around it, but not the legend or numbers on the sides.

**Comment by snibgo** [comment, marked as answer]
Okay. The image has no red pixels. We flood from top-left with red (and a fuzz), then trim all the red away. Windows BAT syntax:
```
magick ^
  OR_Brookings_20200316_TM_geo.tif[0] ^
  -fill Red -fuzz 2%% -draw "color 0,0 floodfill" ^
  +fuzz ^
  -background Red ^
  -define trim:percent-background=0%% -trim +repage ^
  x3.tiff
```
![x](https://user-images.githubusercontent.com/33812368/184573017-554c4946-f77a-4171-8f4d-66d80e381d64.jpg)
(I have uploaded a JPEG version for bandwidth reasons.)

If we are using HDRI, we can use an out-of-range colour such as `sRGB(200%,0,0)` instead of `red`.

**Reply by kj7rrv** [reply]
Thank you! That's exactly what I needed.

**Reply by kj7rrv** [reply]
I switched to using black as the color and 5% fuzz to make sure the entire border is gone, and it's working great. Thank you!
